libs/clients/middlewares/src/lib/enhanced-fetch.ts

```diff
@@ -0,0 +1,184 @@
+import CircuitBreaker from 'opossum'
+import fetch from 'node-fetch'
+import { Logger } from 'winston'
+import { logger as defaultLogger } from '@island.is/logging'
+
+type FetchAPI = WindowOrWorkerGlobalScope['fetch']
+
+interface FetchProblem {
+  type: string
+  title: string
+  status?: number
+  detail?: string
+  instance?: string
+  [key: string]: unknown
+}
+
+interface FetchError extends Error {
+  url: string
+  status: number
+  headers: Headers
+  statusText: string
+  response: Response
+  body?: unknown
+  problem?: FetchProblem
+}
+
+export interface EnhancedFetchOptions {
+  // The name of this fetch function, used in logs and opossum stats.
+  name: string
+
+  // Timeout for requests. Defaults to 10000ms. Can be disabled by passing false.
+  timeout?: number | false
+
+  // Shortcut to disable circuit breaker. Keeps timeout and logging.
+  enableCircuitBreaker?: boolean
+
+  // By default 400 responses are considered warnings and will not open the circuit.
+  // This can be changed by passing `treat400ResponsesAsErrors: true`.
+  // Either way they will be logged and thrown.
+  treat400ResponsesAsErrors?: boolean
+
+  // If true, will log error response body. Defaults to false.
+  // Should only be used if error objects do not have sensitive information or PII.
+  logErrorResponseBody?: boolean
+
+  // Configure circuit breaker logic.
+  opossum?: CircuitBreaker.Options
+
+  // Override logger.
+  logger?: Logger
+
+  // Override fetch function.
+  fetch?: FetchAPI
+}
+
+const createResponseError = async (response: Response, includeBody = false) => {
+  const error = new Error(
+    `Request failed with status code ${response.status}`,
+  ) as FetchError
+  error.name = 'FetchError'
+  const { url, status, headers, statusText } = response
+  Object.assign(error, { url, status, headers, statusText, response })
+
+  const contentType = response.headers.get('content-type')
+  const isJson = contentType === 'application/json'
+  const isProblem = contentType === 'application/problem+json'
+  const shouldIncludeBody = includeBody && (isJson || isProblem)
+  if (isProblem || shouldIncludeBody) {
+    const body = await response.clone().json()
+    if (shouldIncludeBody) {
+      error.body = body
+    }
+    if (isProblem) {
+      error.problem = body
+    }
+  } else if (includeBody) {
+    error.body = await response.clone().text()
+  }
+
+  return error
+}
+
+/**
+ * Creates a fetch function for resilient ops:
+ *
+ * - Includes circuit breaker logic. By default, if more than 50% of the
+ *   requests from the last 10 seconds are misbehaving, we'll open the circuit.
+ *   All future requests will be stopped to lower pressure on the remote server.
+ *   Every 30 seconds we'll allow one request through. If it's successful, we'll
```

Comment:
Curious why every 30 seconds is used?
Is 30 seconds some "known" sweet spot to handle a circuit breaker?

Reply:
I decided to keep (and document) all of opossum's default options (except volumeThreshold), since we don't have the experience to know better defaults.